src/DistributionCalc.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-FileCopyrightText: © 2017, 2018, 2019 dbrock, rain, mrchico
+// SPDX-FileCopyrightText: © 2023 SDAO Foundation <www.sdaofoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity 0.8.19;
+
+interface DistributionCalc {
+    function getAmount(
+        uint256 when,
+        uint256 prev,
+        uint256 tot,
+        uint256 fin,
+        uint256 clf
+    ) external pure returns (uint256);
+}
+
+contract LinearIncreasingDistribution is DistributionCalc {
+    function getAmount(
+        uint256 when,
+        uint256 prev,
+        uint256 tot,
+        uint256 fin,
+        uint256 clf
+    ) external pure returns (uint256) {
+        return (2 * tot * (when - prev)) / (fin ** 2 - clf ** 2);
+    }
+}
+
+contract ConstantDistribution is DistributionCalc {
+    function getAmount(
+        uint256 when,
+        uint256 prev,
+        uint256 tot,
+        uint256 fin,
+        uint256 clf
+    ) external pure returns (uint256) {
+        return (tot * (when - prev)) / (fin - clf);
+    }
+}
```

Comment:
I understand this, however there's a case where it wouldn't work:

When `clf > bng`, the total amount accrued up until the cliff would be disbursed right the way and from what I got from the requirements, this is something that we want to avoid.

Maybe for this specific use case we can assume `eta` will always be `0`, so yeah, this calc might be superfluous, but we should then require `clf == bgn` for it not to break.

However, I believe it would be a good idea to keep the flexibility for consistency and reusability.

---

Regarding changing the structure, I wouldn't mind, but I think there might an upside of preferring Composition to Inheritance here.

Originally, the stream rate for governance tokens would grow from 100k/year to 2MM, then it should remain constant as 4MM.

If we go with inheritance, we would need to replace the `RewardsDistribution` instance.

`distribute()` is probably going to be called by a keeper (be it a bot or in the Keep3r Network). This would mean that we'd have to add the new `RewardsDistribution` address and drop the old one.

If we keep the composition, the only thing we need to change is the `DistributionCalc`, so the `RewardsDistribution` address will remain the same, requiring no changes to the keepers.

Reply:
So we should not be over-engineering this to be the most flexible. The requirements are that we need two types of streams:

 * One that is constant the whole way through and does not use a cliff. So having that `require(clf == bgn)` as you say is the way I would go.
 * The other is one that starts at 0, ramps up to some number and stays constant from there on out. The `RewardDistribution` for that one should not have to be replaced and should automatically switch from linear increasing to constant distribution at the kink point.

Because this is all well-defined, we should use inheritance with a formula to determine the amount to send which matches the project's exact requirements. This is a case where you are tempted to make things ultra-flexible as an engineer (I do this too sometimes), but you need to try to avoid this instinct for SC dev. Mistakes are made usually when trying to go beyond what the exact requirements are. Simple is better.

Reply:
> So we should not be over-engineering this to be the most flexible.

I beg to disagree this is over-engineering, it's a clear case of [a well known design pattern](https://en.wikipedia.org/wiki/Strategy_pattern), which is even [used across the MCD protocol](https://github.com/makerdao/dss/blob/master/src/abaci.sol).

I'm trying to optimize not only for development/deployment cost and complexity, but also for operational cost and complexity.

The solution I'm proposing may not be optimal from the former, but when combining it with the latter, I believe it reaches an optimal point, given the requirements presented so far.

In the end, it's one additional stateless read-only contract being deployed and provided as a constructor parameter. An operational error here would have the effect of distributing the wrong amount of rewards, but only until a new, correct contract is deployed. The risk seems pretty low.

> This is a case where you are tempted to make things ultra-flexible as an engineer (I do this too sometimes), but you need to try to avoid this instinct for SC dev. Mistakes are made when trying to go beyond what the exact requirements are. Simple is better.

We're in an environment of high uncertainty right now, and requirements change way too often. I am not deliberately trying to make things "ultra-flexible", but I believe some degree of flexibility is desired in the current context, otherwise we risk having to come back to the drawing board if requirements change, on a tighter schedule.

"Simple is better [than complex]", but also ["beautiful is better than ugly" and "explicit is better than implicit"](https://peps.python.org/pep-0020/).


> - The other is one that starts at 0, ramps up to some number and stays constant from there on out.

I thought the same initially, but from the last meeting we had, one of the rewards distribution streams would start at 100k/year and end at 2MM/year. After that, it should remain constant as 4MM/year. Please correct me if I'm wrong.

> The RewardDistribution for that one should not have to be replaced and should automatically switch from linear increasing to constant distribution at the kink point.

The way it's currently implemented, the vesting stream is king. The `RerwardsDistribution` contract ensures the full amount being vested ends up in the `StakingRewards` contract, no matter what. In other words, the vesting stream is the single source of truth. There's a clear way to implement the requirements:

1. Start with a vesting stream for the ramp up period and point the `RewardsDistribution` to it.
2. When we are ready to make the switch, create another vesting stream and update `RewardsDistribution` to account for the new stream and the new calc strategy.

The way you are imagining it, `Go from 100k to 2MM and then stay constant at 4MM after X happens` is a little bit harder to model with 1 contract + inheritance, since a `minRate` and a `maxRate` wouldn't suffice. We would have to have a third `rateAfterRampUp` variable.

Another issue is that we don't know exactly where in time the kink point should be, because it depends on a certain smart contract module being available, not a hard date AFAIK. This could be parametrizable, but that could bring even more operational complexity.

If we start to put more control into the `RewardsDistribution` contract itself, we'll need to ensure that either the vesting stream will comply with the requirements (i.e.: will it be able to distribute 4MM/year after kink point?), making the code more rigid and more complex.

The alternative would be to deploy a new base `RewardsDistribution`, but then we would have a different address, requiring any existing automation to update the reference to it.

In either case, there would be more moving parts.

Reply:
Yes you're right it looks like the requirements have changed slightly with 100k start. Nevertheless I would like to pursue an inheritance + put all the custom ramp-up logic inside `RewardsDistribution`. `dss-vest` streams should just be the amount once the ramp-up period ends. The abacus pattern I am against in both the way it was used in the clipper as well as this instance. Sometimes it can be useful, but not in these cases. If requirements change in the next few months we will just need to adjust the code, but probably only slightly.